### PR TITLE
feat: connect to local manifest & copy manifest.json and lu files to current bot

### LINKED ISSUE #7852: Add skill via manifest fails to load remote LU
## Describe the bug

While connecting to a remote skill, the step that is supposed to import the remote LU files fails an error that says "get remote file fail" and requires me to put in my own LU for the skill trigger

Screen Shot 2021-05-17 at 3 45 00 PM

## Version

2.0.0-test

### LINKED ISSUE #7864: Load skill manifests from any location
In composer 2.x a user can load a manifest from an http or https endpoint only, but this should be the only way. 

Manifests describe the skill operations but they don't necessarily need to be exposed on an http endpoint, we do this in samples for convenience but a skill provider may choose not to do this and send the manifest and related files to the consumer offband. In this case, I should be able to load the manifest from the file system if I need to. 

image

Also, I think manifests and related lu files should be stored locally with the bot project once the skill is added and only be rechecked when the developer chooses to (sort of like a decision to update nuget packages). Right now, composer tries to retrieve the manifest every time I open the project and that shouldn't be necessary once a skill has been added and a copy of the manifest and lu files are stored with the botproj.

image

### PR DESCRIPTION
**Pull request by alanlong9278**
## Description

 - connect to local manifest @VanyLaw 
 
 - copy manifest.json and lu files to current bot
 1. from http url
 files

2. from local zip
skillZip

## Task Item

close #7864
close #7852

### CODE REVIEW

**Comment by cwhitten** [comment]
@alanlong9278 does this support the format that PVA will export? It is a .zip format with both the manifest and the language data inside of it.

cc @luhan2017 @gabog

**Comment by alanlong9278** [comment]
> @alanlong9278 does this support the format that PVA will export? It is a .zip format with both the manifest and the language data inside of it.
> 
> cc @luhan2017 @gabog

@cwhitten We will support zip format, I changed  this pr into draft

**Comment by luhan2017** [comment]
> > @alanlong9278 does this support the format that PVA will export? It is a .zip format with both the manifest and the language data inside of it.
> > cc @luhan2017 @gabog
> 
> @cwhitten We will support zip format, I changed this pr into draft

Not yet, we just got the requirement yesterday, working on that.

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx [hunk 1]
```diff
@@ -180,21 +196,54 @@ export const CreateSkillModal: React.FC<CreateSkillModalProps> = (props) => {
...
+        zipContent[fPath] = await files[fPath].async('string');
+        // eslint-disable-next-line no-useless-escape
+        if (fPath.match(/\.([^\.]+)$/)?.[1] === 'json' && isManifestJson(zipContent[fPath])) {
+          manifestFiles.push(files[fPath]);
+          setManifestDirPath(fPath.substr(0, fPath.lastIndexOf('/') + 1));
+        }
+      }
+
+      // update content for detail panel and show it
+      if (manifestFiles.length > 1) {
```
Comment (luhan2017) [inline Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:228]: image

Add a condition for no manifest.json?

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx [hunk 2]
```diff
@@ -73,7 +79,46 @@ export const validateManifestUrl = ({ formData, formDataErrors, setFormDataError
...
+      // eslint-disable-next-line no-useless-escape
+      if (fPath.match(/\.([^\.]+)$/)?.[1] === 'json' && isManifestJson(zipContent[fPath])) {
+        manifestFiles.push(files[fPath]);
+        result.path = fPath.substr(0, fPath.lastIndexOf('/') + 1);
+      }
+    }
+
+    // update content for detail panel and show it
+    if (manifestFiles.length > 1) {
+      result.error = { manifestUrl: formatMessage('zip folder has multiply manifest json') };
```
Comment (EricDahlvang) [inline Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:102]:
```suggestion
      result.error = { manifestUrl: formatMessage('zip folder has multiple manifest json') };
```

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx [hunk 3]
```diff
@@ -73,7 +79,46 @@ export const validateManifestUrl = ({ formData, formDataErrors, setFormDataError
...
+      zipContent[fPath] = await files[fPath].async('string');
+      // eslint-disable-next-line no-useless-escape
+      if (fPath.match(/\.([^\.]+)$/)?.[1] === 'json' && isManifestJson(zipContent[fPath])) {
+        manifestFiles.push(files[fPath]);
+        result.path = fPath.substr(0, fPath.lastIndexOf('/') + 1);
+      }
+    }
+
+    // update content for detail panel and show it
+    if (manifestFiles.length > 1) {
```
Comment (EricDahlvang) [inline Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:101]: The .zip file will also contain the .lu file.

**Comment by luhan2017** [comment]
@alanlong9278 can we add the telemetry data to tell if the manifest is http-based or load from zip content?

**Comment by alanlong9278** [comment]
> @alanlong9278 can we add the telemetry data to tell if the manifest is http-based or load from zip content?

Sure

**Review by luhan2017 — APPROVED** [review]

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx (hunk 3 above)
Comment (luhan2017) [reply Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:101]: @EricDahlvang Here we only check if manifest.json is existed and valid, for lu files, we will check it later and show a warning if it is not accessible like the following, which means, if there is no lu files in .zip, user can still add the manifest but they need to configure their own trigger phrases. 

image

File: Composer/packages/client/src/components/AddRemoteSkillModal/BrowserModal.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    if (zipFile) {
+      // create zip instance
+      const jszip = new JSZip();
+      jszip
+        .loadAsync(zipFile)
+        .then((zip) => {
+          props.onUpdate(zipFile.name, zip.files);
+        })
+        .catch((error) => {
+          console.log(error);
```
Comment (lei9444) [inline Composer/packages/client/src/components/AddRemoteSkillModal/BrowserModal.tsx:27]: Don't use console.log here.

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx [hunk 5]
```diff
@@ -82,12 +127,14 @@ export const getSkillManifest = async (projectId: string, manifestUrl: string, s
     });
     setSkillManifest(data);
   } catch (error) {
+    console.log(error);
```
Comment (lei9444) [inline Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:130]: We should remove this one.

File: Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx [hunk 6]
```diff
@@ -310,14 +383,18 @@ export const CreateSkillModal: React.FC<CreateSkillModalProps> = (props) => {
...
               <div style={{ width: '50%' }}>
-                <TextField
-                  required
-                  errorMessage={formDataErrors.manifestUrl}
-                  label={formatMessage('Skill Manifest URL')}
-                  placeholder={formatMessage('Ask the skill owner for the URL and provide your bot’s App ID')}
-                  value={formData.manifestUrl || ''}
-                  onChange={handleManifestUrlChange}
-                />
+                <div style={{ display: 'flex' }}>
```
Comment (lei9444) [inline Composer/packages/client/src/components/AddRemoteSkillModal/CreateSkillModal.tsx:386]: can we use @emotion/core here?

File: Composer/packages/server/src/controllers/project.ts [hunk 7]
```diff
@@ -351,6 +352,36 @@ async function getSkill(req: Request, res: Response) {
...
+  }
+}
+
+async function removeSkillFiles(req: Request, res: Response) {
+  const projectId = req.params.projectId;
+  const user = await ExtensionContext.getUserFromRequest(req);
+
+  const currentProject = await BotProjectService.getProjectById(projectId, user);
+  if (currentProject !== undefined) {
+    console.log(req.params.name);
```
Comment (lei9444) [inline Composer/packages/server/src/controllers/project.ts:377]: can we remove it?

**Pull request merged**